mlir/lib/Catalyst/Transforms/catalyst_to_llvm.cpp

```diff
@@ -232,6 +263,174 @@ struct PrintOpPattern : public OpConversionPattern<PrintOp> {
     }
 };
 
+// Encodes memref as LLVM struct value:
+//
+// {
+//    i64 rank,
+//    void* data,
+//    i8 type_encoding
+// }
+Value EncodeDataMemRef(Location loc, PatternRewriter &rewriter, MemRefType memrefType,
+                       Type llvmMemrefType, Value memrefLlvm)
+{
+    auto ctx = rewriter.getContext();
+
+    // Encoded memref type: !llvm.struct<(i64, ptr<i8>, i8)>.
+    Type i8 = rewriter.getI8Type();
+    Type i64 = rewriter.getI64Type();
+    Type ptr = LLVM::LLVMPointerType::get(ctx);
+    auto type = LLVM::LLVMStructType::getLiteral(ctx, {i64, ptr, i8});
+
+    std::optional<int8_t> elementDtype = encodeNumericType(memrefType.getElementType());
+
+    // Create values for filling encoded memref struct.
+    Value dtype =
+        rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI8IntegerAttr(elementDtype.value()));
+    Value rank =
+        rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI64IntegerAttr(memrefType.getRank()));
+
+    // Construct encoded memref value.
+    Value memref = rewriter.create<LLVM::UndefOp>(loc, type);
+    // Rank
+    memref = rewriter.create<LLVM::InsertValueOp>(loc, memref, rank, 0);
+
+    // Memref data
+    MemRefDescriptor desc = MemRefDescriptor(memrefLlvm);
+    Value c0 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI64IntegerAttr(0));
+    Value data = rewriter.create<LLVM::GEPOp>(loc, ptr, memrefType.getElementType(),
+                                              desc.alignedPtr(rewriter, loc), c0);
+    memref = rewriter.create<LLVM::InsertValueOp>(loc, memref, data, 1);
+
+    // Dtype
+    memref = rewriter.create<LLVM::InsertValueOp>(loc, memref, dtype, 2);
+
+    return memref;
+}
+
+struct CustomCallOpPattern : public OpConversionPattern<CustomCallOp> {
+    using OpConversionPattern::OpConversionPattern;
+
+    LogicalResult matchAndRewrite(CustomCallOp op, CustomCallOpAdaptor adaptor,
+                                  ConversionPatternRewriter &rewriter) const override
+    {
+        MLIRContext *ctx = op.getContext();
+        Location loc = op.getLoc();
+        // Create function
+        Type ptr = LLVM::LLVMPointerType::get(ctx);
+
+        Type voidType = LLVM::LLVMVoidType::get(ctx);
+        auto type = LLVM::LLVMFunctionType::get(voidType, {/*args=*/ptr, /*rets=*/ptr});
+        auto point = rewriter.saveInsertionPoint();
+        ModuleOp mod = op->getParentOfType<ModuleOp>();
+        rewriter.setInsertionPointToStart(mod.getBody());
+        LLVM::LLVMFuncOp customCallFnOp =
+            rewriter.create<LLVM::LLVMFuncOp>(loc, op.getCallTargetName(), type);
+        customCallFnOp.setPrivate();
+        rewriter.restoreInsertionPoint(point);
```

Comment:
This is optional, but insertion points can also automatically be restored by using a code block, which also helps to visually distinguish modifications in different IR places.
```suggestion
        {
            PatternRewriter::InsertionGuard insertGuard(rewriter);
            ModuleOp mod = op->getParentOfType<ModuleOp>();
            rewriter.setInsertionPointToStart(mod.getBody());
            LLVM::LLVMFuncOp customCallFnOp =
                rewriter.create<LLVM::LLVMFuncOp>(loc, op.getCallTargetName(), type);
            customCallFnOp.setPrivate();
        }
```

Reply:
Thanks a lot I did not know that!

Reply:
It does not work because custom call op is used at the end, if you add the code block the op appears as undefined.

Reply:
Oh my bad, all good then. I guess the op would need to be declared outside the scope but it's not a big change either way, just thought I'd mention it :)